### PR TITLE
[AMDGPU] Prevent folding of the negative i32 literals as i64

### PR DESCRIPTION
**Pull request by rampitec**
We can use sign extended 64-bit literals, but only for signed operands. At the moment we do not know if an operand is signed. Such operand will be encoded as its low 32 bits and then either correctly sign extended or incorrectly zero extended by HW.

### CODE REVIEW

**Comment by rampitec** [comment]
In a long term I think we need to replace INT64 operands with SINT64 and UINT64. That said there are no MVT::s64 and MVT::u64 to use in an instruction profile.

**Comment by arsenm** [comment]
> In a long term I think we need to replace INT64 operands with SINT64 and UINT64. That said there are no MVT::s64 and MVT::u64 to use in an instruction profile.

You would have to bundle that somehow in the profile.

File: llvm/test/CodeGen/AMDGPU/folding-of-i32-as-i64.mir [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+body: |
+  bb.0:
+    ; GCN-LABEL: name: imm64_shift_int32_const
+    ; GCN: [[S_MOV_B:%[0-9]+]]:sreg_64 = S_MOV_B64_IMM_PSEUDO -2147483648
+    ; GCN-NEXT: [[S_LSHL_B64_:%[0-9]+]]:sreg_64 = S_LSHL_B64 [[S_MOV_B]], 1, implicit-def $scc
+    ; GCN-NEXT: S_ENDPGM 0, implicit [[S_LSHL_B64_]]
+    %0:sreg_64 = S_MOV_B64_IMM_PSEUDO 18446744071562067968
+    %1:sreg_64 = S_LSHL_B64 %0, 1, implicit-def $scc
+    S_ENDPGM 0, implicit %1
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/folding-of-i32-as-i64.mir:19]: Needs some negative tests. Can you also add an end to end IR test that hits this?

**Review by arsenm — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 2]
```diff
@@ -5500,6 +5500,15 @@ bool SIInstrInfo::isOperandLegal(const MachineInstr &MI, unsigned OpIdx,
     if (Is64BitOp && !AMDGPU::isValid32BitLiteral(Imm, Is64BitFPOp) &&
         !AMDGPU::isInlinableLiteral64(Imm, ST.hasInv2PiInlineImm()))
       return false;
+
+    // FIXME: We can use sign extended 64-bit literals, but only for signed
+    //        operands. At the moment we do not know if an operand is signed.
+    //        Such operand will be encoded as its low 32 bits and then either
+    //        correctly sign extended or incorrectly zero extended by HW.
+    if (Is64BitOp && !Is64BitFPOp && isInt<32>(Imm) &&
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:5508]: I don't understand what the `isInt<32>` check is for. This should also apply to the `isUInt<32>` case, shouldn't it? So just remove that part?

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 3]
```diff
@@ -5500,6 +5500,15 @@ bool SIInstrInfo::isOperandLegal(const MachineInstr &MI, unsigned OpIdx,
...
+        (int32_t)Lo_32(Imm) < 0 &&
+        !AMDGPU::isInlinableLiteral64(Imm, ST.hasInv2PiInlineImm()))
+      return false;
   }
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:5512]: Could refactor this whole block as:
```
  if (Is64BitOp && !AMDGPU::isInlinableLiteral64(Imm, ST.hasInv2PiInlineImm())) {
    if (!AMDGPU::isValid32BitLiteral(Imm, Is64BitFPOp))
      return false;
    ... new code here ...
  }
```

**Comment by rampitec** [comment]
> > In a long term I think we need to replace INT64 operands with SINT64 and UINT64. That said there are no MVT::s64 and MVT::u64 to use in an instruction profile.
> 
> You would have to bundle that somehow in the profile.

Right, VT to use in profile is also missing.

File: llvm/test/CodeGen/AMDGPU/folding-of-i32-as-i64.mir (hunk 1 above)
Comment (rampitec) [reply llvm/test/CodeGen/AMDGPU/folding-of-i32-as-i64.mir:19]: I will add negative (or I guess positive) tests here, but for the end-to-end it is not currently possible. It needs us to select this mov, which we do not do yet. The point of this patch series is to prepare to that selection, and that is how I hit the problem - I have a patch to select it and fixing problems which will arise.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 2 above)
Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:5508]: This check bails on 0xffffffff for instance. It I remove it this folding in the fold-short-64-bit-literals.mir will not happen:
```
     ; GCN-LABEL: name: fold_uint_32bit_literal_sgpr
     ; GCN: [[DEF:%[0-9]+]]:sreg_64 = IMPLICIT_DEF
-    ; GCN-NEXT: [[S_AND_B64_:%[0-9]+]]:sreg_64 = S_AND_B64 [[DEF]], 4294967295, implicit-def $scc
+    ; GCN-NEXT: [[S_MOV_B64_:%[0-9]+]]:sreg_64 = S_MOV_B64 4294967295
+    ; GCN-NEXT: [[S_AND_B64_:%[0-9]+]]:sreg_64 = S_AND_B64 [[DEF]], [[S_MOV_B64_]], implicit-def $scc
```

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 3 above)
Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Done

File: llvm/test/CodeGen/AMDGPU/folding-of-i32-as-i64.mir (hunk 1 above)
Comment (rampitec) [reply llvm/test/CodeGen/AMDGPU/folding-of-i32-as-i64.mir:19]: Added more positive and negative tests.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 2 above)
Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Well, I guess this folding is illegal too:
```
    %0:sreg_64 = S_MOV_B64_IMM_PSEUDO 4294967295
    %1:sreg_64 = S_ASHR_I64 %0, 1, implicit-def $scc
```
Quite unfortunately it is impossible to distinguish I32 and B32 operands here.

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Done.

File: llvm/test/CodeGen/AMDGPU/folding-of-i32-as-i64.mir (hunk 1 above)
Comment (rampitec) [reply llvm/test/CodeGen/AMDGPU/folding-of-i32-as-i64.mir:19]: @arsenm is test look good for you now?

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 4]
```diff
@@ -5497,9 +5497,18 @@ bool SIInstrInfo::isOperandLegal(const MachineInstr &MI, unsigned OpIdx,
...
+    if (Is64BitOp &&
+        !AMDGPU::isInlinableLiteral64(Imm, ST.hasInv2PiInlineImm())) {
+      if (!AMDGPU::isValid32BitLiteral(Imm, Is64BitFPOp))
+        return false;
+
+      // FIXME: We can use sign extended 64-bit literals, but only for signed
+      //        operands. At the moment we do not know if an operand is signed.
+      //        Such operand will be encoded as its low 32 bits and then either
+      //        correctly sign extended or incorrectly zero extended by HW.
+      if (!Is64BitFPOp && (int32_t)Lo_32(Imm) < 0)
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:5509]: Nit: `(int32_t)Imm` is simpler.

**Review by jayfoad — APPROVED** [review]
LGTM if Matt is happy with the tests.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 4 above)
Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Done.

**Review by arsenm — APPROVED** [review]

**Pull request merged**